### PR TITLE
Adding Lex v2 event and response

### PR DESCRIPTION
**Pull request by Sordie**
*Issue #, if available:*
#242

*Description of changes:*
Adding both Lex version 2 event and response as defined here: https://docs.aws.amazon.com/lexv2/latest/dg/lambda.html

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    @Data
+    @Builder(setterPrefix = "with")
+    @NoArgsConstructor
+    @AllArgsConstructor
+    public static class Transcription implements Serializable, Cloneable {
+        private String transcription;
+        private TranscriptionConfidence transcriptionConfidence;
+        private ResolvedContext resolvedContext;
+        private Map<String, Slot> ResolvedSlots;
```
Comment (msailes) [inline aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java:210]: is this a typo or does it only work with Capital 'R'?

**Comment by msailes** [comment]
Could you add a test into the serialization module?

https://github.com/aws/aws-lambda-java-libs/blob/master/aws-lambda-java-serialization/src/test/java/com/amazonaws/services/lambda/runtime/serialization/events/LambdaEventSerializersTest.java

File: aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java (hunk 1 above)
Comment (Sordie) [reply aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java]: Yes, must have missed that typo. Thanks for pointing it out.

File: aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+    @Data
+    @Builder(setterPrefix = "with")
+    @NoArgsConstructor
+    @AllArgsConstructor
+    public static class SentimentScore implements Serializable, Cloneable {
+        private double mixed;
+        private double negative;
+        private double neutral;
+        private double positiv;
```
Comment (msailes) [inline aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java:126]: is this a typo?

**Comment by msailes** [comment]
Hi @Sordie,

Could this PR be change to target the [V4](https://github.com/aws/aws-lambda-java-libs/tree/events-v4-serialization-v2) branch? You can use the Jackson annotations to handle the capitalisation.

**Comment by msailes** [comment]
Could you check the tests?

**Comment by Sordie** [comment]
Hi @msailes,
I have added an additional Lex event that includes the optional parts for both sentiment and Kendra. However, I’m still struggling with the tests. As already mentioned above the tests currently fail, because the slots map is empty after the deserialization and subsequent serialization. At this point I haven’t found an easy way to fix this. Could you share your opinion on how to best proceed? Maybe you already have an idea how to solve this issue?

**Comment by msailes** [comment]
The `testLexV2Event` seems to be failing because the test input `lex_v2_event.json` contains a Map with null values. In the `JacksonPojoSerializer` class we set `mapper.setSerializationInclusion(Include.NON_NULL);` which excludes null values.

```
          "intent": {
              "slots": {
                  "ReturnDate": null,
                  "PickUpDate": null,
                  "DriverAge": null,
                  "CarType": null,
                  "PickUpCity": null
              },
```

I don't know if this is bad data or possible? @Sordie do you know?

**Comment by msailes** [comment]
Same for the `testLexV2KendraSentimentEvent` test.

**Comment by Sordie** [comment]
@msailes thanks for pointing out where the configuration to exclude null values is, as I had not seen it before.
Regarding the possibility of bad data in the test events, both event json files contain data taken directly from the input of a TypeScript Lambda that was attached to a Lex bot. I only replaced some unique identifiers with example strings. Therefore, I would say that this data is representative of the actual input a lambda will receive from the Lex service.

**Comment by msailes** [comment]
@Sordie thank you.

I think the tests you've added should be slightly changed then to show this behavior. The actual will be different to the json from the test file because maps which include items will null values are excluded. 

@andclt, @smirnoal I'm not sure how you want this implementing, maybe two files would be a suitable way?

@andclt, @smirnoal I found this a hard problem to track down since there was no specific JSON knowledge in the testing lib. I would suggest with move to [JSONassert](https://github.com/skyscreamer/JSONassert) either in this PR or another seperate one. This library gave much better error messages and I was able to spot the problem straight away.

**Comment by andclt** [comment]
First of all thanks for contributing @Sordie! 

I agree with @msailes' suggestion. However, if you want to keep those null values after the serialization, I think you can achieve it adding the `@JsonInclude` annotation (https://fasterxml.github.io/jackson-annotations/javadoc/2.9/com/fasterxml/jackson/annotation/JsonInclude.html) at the field level. This will override the global custom serialization behaviour that is currently excluding null values.

File: aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+ *
+ * @see <a href=
+ *      "https://docs.aws.amazon.com/lexv2/latest/dg/lambda.html#lambda-input-format">Using
+ *      an AWS Lambda function</a>
+ */
+@Data
+@Builder(setterPrefix = "with")
+@NoArgsConstructor
+@AllArgsConstructor
+public class LexV2Event implements Serializable, Cloneable {
```
[the remark this answers is not in the document]
Comment (Sordie) [reply aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java]: Ok, I have removed the cloneable interface from all classes I created

File: aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Sordie) [reply aws-lambda-java-events/src/main/java/com/amazonaws/services/lambda/runtime/events/LexV2Event.java]: Makes sense, I have updated the year to 2022 :D

**Comment by andclt** [comment]
Hi @Sordie,

As I said in a previous message on this thread, I think we should use the `@JsonInclude` annotation to override our current  Jackson configuration (that's excluding null values) instead of using a different json file to compare the lex event v2 serialization.

I was able to test successfully the following implementation, could you please take a look at it?

```
private Map<String, Slot> slots;

@JsonInclude(value=JsonInclude.Include.NON_NULL, content=JsonInclude.Include.ALWAYS)
public Map<String, Slot> getSlots() {
    return this.slots;
}

@JsonInclude(value=JsonInclude.Include.NON_NULL, content=JsonInclude.Include.ALWAYS)
public void setSlots(Map<String, Slot> slots) {
    this.slots = slots;
}
```

**Comment by Sordie** [comment]
Hi @andclt, thanks for providing a code snippet. I will add your suggestion next week once I have a computer in front of me again.

**Comment by Sordie** [comment]
Hi @andclt, I have applied your suggestion and removed the now unused json files.

**Comment by andclt** [comment]
Hi @Sordie, thank you! Approved!

**Review by andclt — APPROVED** [review]

**Pull request merged**